iothub_client/samples/iothub_client_device_cbor_sample/iothub_client_device_cbor_sample.c

```diff
@@ -0,0 +1,956 @@
+    }
+    else
+    {
+        // The content type system property value is application-defined and must reflect how the
+        // service application has chosen to define it.
+        if (strcmp(content_type, CONTENT_TYPE_C2D) == 0)
+        {
+            (void)printf("The device client received expected system property content type value: %s.\n", CONTENT_TYPE_C2D);
+
+            // This content-type refers to how the message was created on the service-client application:
+            // either from a byte array or a string. This sample assumes a byte array was used.
+            IOTHUBMESSAGE_CONTENT_TYPE message_content_type = IoTHubMessage_GetContentType(message);
+            if (message_content_type == IOTHUBMESSAGE_BYTEARRAY)
+            {
+                const unsigned char* payload;
+                size_t payload_length;
+
+                if (IoTHubMessage_GetByteArray(message, &payload, &payload_length) != IOTHUB_MESSAGE_OK)
+                {
+                    (void)printf("Failed to retrieve message as a byte array.\n");
+                }
+                else
+                {
+                    (void)printf("The device client retrieved the message as a byte array.\n");
+                    // The application should parse the payload as the expected content type.
+                }
+            }
+        }
+        else
+        {
+            (void)printf("The device client did not receive expected system property content type value: %s.\n", CONTENT_TYPE_C2D);
+        }
+    }
+
+    (void)printf("\n"); // Formatting
+
+    return IOTHUBMESSAGE_ACCEPTED;
+}
+
+//
+// Encoding/Decoding with CBOR library
+//
+static bool parse_cbor_desired_property(DEVICE_TWIN_UPDATE_STATE update_state, const unsigned char* payload, Car* out_car)
+{
+    CborError rc; // CborNoError == 0
+    bool result;
+
+    CborParser parser;
+    CborValue root;
+    CborValue state_root;
+
+    CborValue change_oil_reminder;
+    CborValue max_speed;
+    CborValue software_version;
+
+    rc = cbor_parser_init(payload, strlen((char*)payload), 0, &parser, &root);
+    if (rc)
+    {
+        (void)printf("Failed to initiate parser: CborError %d.\n", rc);
+        exit(rc);
+    }
+
+    // Check if this is the full twin document or only the desired properties.
+    if (update_state == DEVICE_TWIN_UPDATE_COMPLETE) // full twin document (desried and reported)
+    {
+        rc = cbor_value_map_find_value(&root, twin_desired_property, &root);
+        if (rc)
+        {
+            (void)printf("Error when searching for %s: CborError %d.\n", twin_desired_property, rc);
+            exit(rc);
+        }
+    }
+
+    // change_oil_reminder
+    rc = cbor_value_map_find_value(&root, twin_property_change_oil_reminder_name, &change_oil_reminder);
+    if (rc)
+    {
+        (void)printf("Error when searching for %s: CborError %d.\n", twin_property_change_oil_reminder_name, rc);
+        exit(rc);
```

Comment:
is this the right behavior? If I start up a new device with no device twin to start and only update one at a time in desired, is this going to `exit` the app?

Reply:
good catch. i'll double check.

Reply:
Ok, so if the parameter is not found, this won't return an error. Instead it will return OK and the type will be defined as INVALID.  That is checked on the line 616.  If its not found, thats when the return is false (meaning the parameter was not found).  I do need to update this logic so not to override one another, and also if no parameters are found, then there is nothing to update.  So I do want to keep that bool present.